Home task3

Peatro opened the pull request:
Добавлены собственные классы LinkdList и ArrayList
Добавлены методы сортировки в соответствии с заданием
Добавлено тестирование классов

Peatro committed:
Добавлена реализация собственного класса ArrayList

Peatro committed:
Добавлена реализация сортировки слиянием в класс CustomArrayList

Peatro committed:
Добавлено тестирование класса CustomArrayList

Peatro committed:
Добавлена реализация собственного класса LinkedList

Peatro committed:
Добавлена реализация быстрой сортировки в классе CustomLinkedList

Peatro committed:
Добавлено тестирование класса CustomLinkedList

VeraKalinkin commented on Homework02/src/main/java/org/example/CustomLinkedList.java:
```diff
@@ -0,0 +1,211 @@
+
+        if (first == null) {
+            first = newNode;
+            last = newNode;
```
можно в одну строку сделать

Peatro replied:
Можно, но для читаемости думаю удобнее так

ktaradanov replied:
Лучше для читабельности оставить, не экономим строчки, они бесплатные)

VeraKalinkin commented on Homework02/src/main/java/org/example/CustomLinkedList.java:
```diff
@@ -0,0 +1,211 @@
+        if (last == null) {
+            System.out.println("List is empty");
+            return null;
+        }
```
можно эту проверку вынести в отдельный метод, она используется несколько раз

Peatro replied:
Если про повтороное использование идёт речь о методе removeFirst(), то там идёт проверка на null поля first

ktaradanov replied:
Проверку одного значения не особо имеет смысл выносить, получится метод ради метода. Другая проблема есть, если вызываем удаление на пустом списке, возвращать null не лучшая идея. Таким образом сажаешь потенциальное NPE, лучше сразу исключение бросай

VeraKalinkin commented on Homework02/src/main/java/org/example/CustomArrayList.java:
```diff
@@ -0,0 +1,141 @@
+        if (size == elements.length) {
+            grow();
+        }
+        System.arraycopy(elements, index, elements, index + 1, size - index);
```
используются методы из коллекций

Peatro replied:
В задании ничего не было про это указано конкретно, если будет проблемой, то будет изменено

ktaradanov replied:
Метод не из коллекций, это метод класса System. Исправлений не требуется
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html

ktaradanov commented on Homework02/src/main/java/org/example/CustomLinkedList.java:
```diff
@@ -0,0 +1,211 @@
+
+        if (first == null) {
+            System.out.println("List is empty");
+            return null;
```
тоже NPE

ktaradanov commented on Homework02/src/test/java/org/example/CustomArrayListTest.java:
```diff
@@ -0,0 +1,180 @@
+    CustomArrayList<Integer> customArrayListInteger = new CustomArrayList<>();
+
+    @Test
+    public void customArrayListConstructorTest() {
```
выброс исключения это отдельный тест кейс, лучше вынести в отдельный тест

ktaradanov replied:
тут и в других местах

ktaradanov commented on Homework02/src/test/java/org/example/CustomArrayListTest.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Test
+    public void customArrayListAddTest() {
```
можно параметризованный тест сделать

ktaradanov replied:
тут и в других местах

ktaradanov commented on Homework02/src/test/java/org/example/CustomArrayListTest.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Test
+    public void customArrayListIsEmptySize() {
```
по названию теста как будто проверяешь size для пустого листа, а в самом тесте он не пустой

ktaradanov commented on Homework02/src/test/java/org/example/CustomArrayListTest.java:
```diff
@@ -0,0 +1,180 @@
+    private static <T extends Comparable<T>> void assertArrayEquals(T[] array, CustomArrayList<T> list) {
+        if (array.length != list.size()) {
+            throw new IllegalStateException("Length should be EQUALS! - Array size: " + array.length + " List size: " + list.size());
+        }
```
пустую строку между конструкциями нужно

ktaradanov replied:
по всему коду проверь форматирование

Peatro committed:
Перелопатил тестовый класс для ArrayList:
- Добавлены параметрезированные тесты
- Тесты на исключения вынесены отдельным кейсом
- Добавлен тест на увеличение размера внутреннего массива, при граничном случае
Изменения в ArrayList:
- Исправлен косяк в методе checkIndex()
- Убраны лишние импорты

Peatro committed:
Переработан класс тестирования LinkedList:
- Добавлены параметризированные тесты
- Исключения обрабатываются отдельно
- Тестами покрыт весь класс
Доработки в LinkedList:
- Исправлен метод добавления значения по индексу, а конкретно linkBefore(). До этого работал не верно на граничных значениях
- Переименован метод add() на addById(), для большего понимания
- Методы removeFirst() и removeLast() теперь бросают NoSuchElementException, если применяются на пустом листе
- Метод getSize() переименован в size() для соответствия

Peatro committed:
fixup! Добавлена строчка между конструкциями, для читаемости
